src/empty.js

```diff
@@ -1,21 +1,38 @@
 var _curry1 = require('./internal/_curry1');
-var _hasMethod = require('./internal/_hasMethod');
 
 
 /**
- * `empty` returns an empty list for any argument, except when the argument satisfies the
- * Fantasy-land Monoid spec. In that case, this function will return the result of invoking
- * `empty` on that Monoid.
+ * Returns the empty value of its argument's type. The argument must be a
+ * [Monoid](https://github.com/fantasyland/fantasy-land#monoid), Array,
+ * Object, or String.
  *
  * @func
  * @memberOf R
  * @category Function
- * @sig * -> []
- * @return {Array} An empty array.
+ * @sig Monoid m => m a -> m a
+ * @param {*} x
+ * @return {*}
  * @example
  *
- *      R.empty([1,2,3,4,5]); //=> []
+ *      R.empty(Just(42));      //=> Nothing()
+ *      R.empty([1, 2, 3]);     //=> []
+ *      R.empty('unicorns');    //=> ''
+ *      R.empty({x: 1, y: 2});  //=> {}
  */
 module.exports = _curry1(function empty(x) {
-  return _hasMethod('empty', x) ? x.empty() : [];
+  var repr = Object.prototype.toString.call(x);
+  return (
+    x != null && typeof x.empty === 'function' ?
+      x.empty() :
+    x != null && typeof x.constructor != null && typeof x.constructor.empty === 'function' ?
+      x.constructor.empty() :
+    repr === '[object Array]' ?
+      [] :
+    repr === '[object Object]' ?
+      {} :
+    repr === '[object String]' ?
+      '' :
+    // else
+      void 0
+  );
 });
```

Comment:
That's a lot of ternary operators!

Do we really want to code in that style?

I'm especially bothered because it forces us to pre-calculate an unnecessary value, `repr`, if either of the first two cases passes.  It's not really a question of performance, but of cleanliness.